matcher.go

```diff
@@ -175,40 +177,144 @@ func (m *Matcher) matchExec(token TokenType, tokenData []byte, node *ExecNode) e
 			}
 
 			if keyElem, ok := node.Elems[string(keyBytes)]; ok {
-				//log.Printf("MATCH ELEM: %s", keyBytes)
+				// Run the execution node that applies to this particular
+				// key of the object.
 				m.matchExec(token, tokenData, keyElem)
+
+				// Check if running this keys execution has resolved the entirety
+				// of the expression, if so we can leave immediately.
+				if m.buckets.IsResolved(0) {
+					return nil
+				}
 			} else {
-				//log.Printf("SKIP ELEM: %s", keyBytes)
+				// If we don't have any parse requirements for this key in
+				// the object, we can just skip its value and continue
 				m.skipValue(token)
 			}
-
-			if m.buckets.IsResolved(0) {
-				return nil
-			}
 		}
-	}
+	} else if token == tknArrayStart {
+		if len(node.Loops) == 0 {
+			// If we have no loop handlers, we can just skip the whole thing...
+			m.skipValue(token)
+		} else {
+			// TODO(brett19): We need to improve this.  The scanner/tokenizer actually
+			// has a bunch of behind-the-scenes state that we need to be cautious of.  In
+			// this case, because we always stop at the edge of a value, process that
+			// value and then return to the begining, the internal stacks are identical
+			// and we don't need to worry about it.  This is extremely dangerous and very
+			// prone to being broken in the future though.  State saving needs to be more
+			// inclusive, or the tokenizer needs to be modified to not need internal state.
+			savePos := m.tokens.pos
+
+			for loopIdx, loop := range node.Loops {
+				if loopIdx != 0 {
+					// If this is not the first loop, we will need to reset back to the
+					// begining of the array the loops are scanning.  In the future, perhaps
+					// we can add support for parallel ExecNode handling and do it in one pass.
+					m.tokens.seek(savePos)
+				}
 
-	if token == tknArrayStart {
-		for i := 0; ; i++ {
-			token, _, err := m.tokens.step()
-			if err != nil {
-				return err
-			}
+				// We need to keep track of the overall loop result value while the bin tree
+				// is being iterated on, reset, etc...
+				var loopState bool
+				if loop.Mode == LoopTypeAny {
```

Comment:
Setting to false is technically not necessary due to default value

Reply:
Whilst you are 100% correct, this was actually a conscious decision to avoid there being any ambiguity or potential to confuse the default value in the future.  If I relied on the default value, potentially new nodeType's could get the wrong default value rather than the panic that we have today.  It may have been worthwhile for me to make it a switch-case with fall-through items to at least make it more clear though.